### PR TITLE
IIS Web App Management Task Phase 2

### PR DESCRIPTION
**Pull request by asranja**

### CODE REVIEW

File: Tasks/IISWebAppManagementOnMachineGroup/task.json [hunk 1]
```diff
@@ -23,59 +23,101 @@
...
-			"groupName": "website",
-			"defaultValue": "false",
-			"helpMarkDown": "Select the option to create a website or to update an existing website."
+			"options": {
+				"IISWebsite": "IIS Website",
+				"IISWebApplication": "IIS Web Application",
+				"IISVirtualDirectory": "IIS Virtual Directory",
+				"IISApplicationPool": "IIS Application Pool"
+			},
+			"helpMarkDown": ""
```
Comment (asranja) [inline Tasks/IISWebAppManagementOnMachineGroup/task.json]: Will update all the help mark downs.

File: Tasks/IISWebAppManagementOnMachineGroup/IISWebAppManagementOnMachineGroup.ps1 [hunk 2]
```diff
@@ -25,42 +25,190 @@ try {
...
+        $appPoolPasswordForWebsite = "$appPoolPasswordForWebsite" | ConvertTo-SecureString -AsPlainText -Force
+        $appPoolCredentials = New-Object System.Management.Automation.PSCredential ("$appPoolUsernameForWebsite", $appPoolPasswordForWebsite)
+    }
+
+    $appCmdCommands = Escape-SpecialChars -str $appCmdCommands
+
+    Execute-Main -ActionIISWebsite $actionIISWebsite -WebsiteName $websiteName -PhysicalPath $websitePhysicalPath -PhysicalPathAuth $websitePhysicalPathAuth -PhysicalPathAuthCredentials $websitePhysicalPathAuthCredentials -AddBinding $addBinding -Protocol $protocol -IpAddress $ipAddress -Port $port -HostName $hostName -ServerNameIndication $serverNameIndication -SslCertThumbPrint $sslCertThumbPrint -CreateAppPool $createOrUpdateAppPoolForWebsite -AppPoolName $appPoolNameForWebsite -DotNetVersion $dotNetVersionForWebsite -PipeLineMode $pipeLineModeForWebsite -AppPoolIdentity $appPoolIdentityForWebsite -AppPoolCredentials $appPoolCredentials -AppCmdCommands $appCmdCommands
+}
+
+function Get-HostName
```
Comment (kmkumaran) [inline Tasks/IISWebAppManagementOnMachineGroup/IISWebAppManagementOnMachineGroup.ps1]: Isn't these helper and part of the PS module? Should we replicate this in all Tasks?

File: Tasks/IISWebAppManagementOnMachineGroup/Utility.ps1 [hunk 3]
```diff
@@ -50,12 +48,12 @@ function Validate-Inputs
...
 
     Write-Verbose "Validating website and application pool inputs"
-    if($createWebsite -ieq "true" -and [string]::IsNullOrWhiteSpace($websiteName))
+    if(($createWebsite -ieq "true" -or $actionIISWebsite -ieq "CreateOrUpdateWebsite") -and [string]::IsNullOrWhiteSpace($websiteName))
     { 
         throw "Website Name cannot be empty if you want to create or update the target website."
     }
 
-    if($createAppPool -ieq "true" -and [string]::IsNullOrWhiteSpace($appPoolName))
+    if(($createAppPool -ieq "true" -or $actionIISApplicationPool -ieq "CreateOrUpdateAppPool") -and [string]::IsNullOrWhiteSpace($appPoolName))
```
Comment (kmkumaran) [inline Tasks/IISWebAppManagementOnMachineGroup/Utility.ps1]: Isn't validation of inputs not added for other two new commands (example: create application?)

File: Tasks/IISWebAppManagementOnMachineGroup/task.json [hunk 4]
```diff
@@ -23,59 +23,101 @@
...
+		},
+		{
+			"name": "ActionIISWebsite",
+			"type": "pickList",
+			"label": "Action",
+			"required": true,
+			"visibleRule": "IISDeploymentType = IISWebsite",
+			"options": {
+				"CreateOrUpdateWebsite": "Create Or Update",
```
Comment (kmkumaran) [inline Tasks/IISWebAppManagementOnMachineGroup/task.json]: Can you introduce the start/Stop actions once the support is added.

File: Tasks/IISWebAppManagementOnMachineGroup/task.json [hunk 5]
```diff
@@ -23,59 +23,101 @@
...
+		},
+		{
+			"name": "ActionIISApplicationPool",
+			"type": "pickList",
+			"label": "Action",
+			"required": true,
+			"visibleRule": "IISDeploymentType = IISApplicationPool",
+			"options": {
+				"CreateOrUpdateAppPool": "Create Or Update",
+				"StartAppPool": "Start App pool",
```
Comment (kmkumaran) [inline Tasks/IISWebAppManagementOnMachineGroup/task.json]: I assume the start/stop of App pool is not available either. Bring it in json once the backend is available.

File: Tasks/IISWebAppManagementOnMachineGroup/task.json [hunk 6]
```diff
@@ -191,28 +232,303 @@
...
 		},
 		{
-			"name": "CreateAppPool",
+			"name": "CreateOrUpdateAppPoolForWebsite",
+			"type": "boolean",
+			"label": "Create or Update App pool",
+			"required": false,
+			"groupName": "Website",
+			"defaultValue": "false",
+			"helpMarkDown": ""
```
Comment (kmkumaran) [inline Tasks/IISWebAppManagementOnMachineGroup/task.json]: Missing helpmarkdown

File: Tasks/IISWebAppManagementOnMachineGroup/task.json [hunk 7]
```diff
@@ -191,28 +232,303 @@
...
+			"visibleRule": "AppPoolIdentityForWebsite = SpecificUser"
+		},	
+		{
+			"name": "ParentWebsiteNameForVD",
+			"type": "string",
+			"label": "Parent Website Name",
+			"defaultValue": "",
+			"required": true,
+			"visibleRule": "IISDeploymentType = IISVirtualDirectory",
+			"helpMarkDown": ""
```
Comment (kmkumaran) [inline Tasks/IISWebAppManagementOnMachineGroup/task.json]: Empty helpmarkdown

File: Tasks/IISWebAppManagementOnMachineGroup/task.json [hunk 8]
```diff
@@ -191,28 +232,303 @@
...
+			"helpMarkDown": "Provide the user name that will be used to access the website's physical path."
+		},
+		{
+			"name": "VDAuthUserPassword",
+			"type": "string",
+			"label": "Password",
+			"required": false,
+			"defaultValue": "",
+			"visibleRule": "VDPhysicalPathAuth = VDWindowsAuth",
+			"helpMarkDown": "Provide the user's password that will be used to access the website's physical path. If you are using a gMSA, this is not required."
```
Comment (kmkumaran) [inline Tasks/IISWebAppManagementOnMachineGroup/task.json]: Is gMSA configuration available for virtual directories?

**Pull request merged**